File: 84/solution_dp.rb
# DP

# @param {Integer[]} heights
# @return {Integer}
def largest_rectangle_area(heights)
    n, result = heights.size, 0
    
    n.times do |i|
        if i + 1 < n && heights[i] <= heights[i + 1]
            next
        end
        
        h = heights[i]
        i.downto(0) do |j|
            h = [h, heights[j]].min
            w = i - j + 1
            result = [result, h * w].max
        end
    end
    
    result
end

File: 84/solution_dp.c
// DP

int min(const int a, const int b) {
    return a < b ? a : b;
}

int max(const int a, const int b) {
    return a > b ? a : b;
}

int largestRectangleArea(int* heights, int heightsSize){
    int n = heightsSize;
    int result = 0;
    
    for(int i = 0; i < n; i++) {
        if(i + 1 < n && heights[i] <= heights[i + 1]) {
            continue;
        }
        
        int h = heights[i], w;
        for(int j = i; j >= 0; j--) {
            h = min(h, heights[j]);
            w = i - j + 1;
            result = max(result, h * w);
        }
    }
    
    return result;
}
